File: stack-easy/496.下一个更大元素-i.java
/*
 * @lc app=leetcode.cn id=496 lang=java
 *
 * [496] 下一个更大元素 I
 */

// @lc code=start

// 15/15 cases passed (5 ms)
// Your runtime beats 82.74 % of java submissions
// Your memory usage beats 19.53 % of java submissions (38.8 MB)
class Solution {
    /* 单调栈 Monotonic stack */
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Deque<Integer> stack = new LinkedList<Integer>();
        // 先对 nums2 中的每一个元素，求出其下一个更大的元素。将这一对值放入哈希映射（HashMap）中
        HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
        for (int i = 0; i < nums2.length; i++) {
            while (!stack.isEmpty() && stack.peek() < nums2[i]) {
                map.put(stack.pop(), nums2[i]);
            }
            stack.push(nums2[i]);
        }
        while (!stack.isEmpty()) {
            map.put(stack.pop(), -1);
        }
        int[] res = new int[nums1.length];
        for (int i = 0; i < nums1.length; i++) {
            res[i] = map.get(nums1[i]);
        }
        return res;
    }
}

// @lc code=end

File: stack-easy/496.下一个更大元素-i.cpp
/*
 * @lc app=leetcode.cn id=496 lang=cpp
 *
 * [496] 下一个更大元素 I
 */

#include <stack>
#include <unordered_map>
#include <vector>

using namespace std;

// @lc code=start

// @monotonic-stack
class Solution_1 {
   public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        stack<int> s;
        //记录每个值对应的下一个最大值
        unordered_map<int, int> map;
        //正序遍历
        for (int i = 0; i < nums2.size(); i++) {
            while (!s.empty() && s.top() < nums2[i]) {
                //更新map
                map[s.top()] = nums2[i];
                s.pop();
            }
            // map初始化
            map[nums2[i]] = -1;
            s.push(nums2[i]);
        }
        //处理返回
        for (int i = 0; i < nums1.size(); i++) {
            nums1[i] = map[nums1[i]];
        }
        return nums1;
    }
};

// @monotonic-stack
// 链接：https://leetcode-cn.com/problems/next-greater-element-i/solution/496xia-yi-ge-geng-da-yuan-su-i-c-dan-dia-37dl/
class Solution_2 {
   public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        int m = nums1.size();
        int n = nums2.size();
        unordered_map<int, int> map;
        stack<int> s;
        // 从后往前
        for (int i = n - 1; i >= 0; i--) {
            // 单调递减栈
            while (!s.empty() && nums2[i] > s.top()) {
                s.pop();
            }
            map[nums2[i]] = s.empty() ? -1 : s.top();
            s.push(nums2[i]);
        }

        vector<int> ans(m);
        for (int i = 0; i < m; i++) {
            ans[i] = map[nums1[i]];
        }

        return ans;
    }
};

// @lc code=end
